src/app/icd/client/ICDClientStorageDelegate.h

```diff
@@ -0,0 +1,94 @@
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "ICDClientInfo.h"
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The ICDClientStorageDelegate class is an abstract interface that defines the operations
+ * for storing, retrieving and deleting ICD client information in persistent storage.
+ */
+class ICDClientStorageDelegate
+{
+public:
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientStorageDelegate() = default;
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
+
+    /**
+     * Called during ICD device registration in commissioning, commissioner/controller
+     * provides raw key data, the shared key handle in clientInfo is updated based upon raw key data
+     *
+     * @param[inout] aICDClientInfo the ICD Client information to be updated with keyData and be saved
+     * @param[in] aKeyData raw key data provided by application
+     */
+    virtual CHIP_ERROR SetKey(ICDClientInfo & aClientInfo, const ByteSpan aKeyData) = 0;
+
+    /**
+     * Store updated ICD ClientInfo to storage when ICD registration completes or check-in message
+     * comes.
+     *
+     * @param[in] aICDClientInfo the updated ICD Client Info.
+     */
+    virtual CHIP_ERROR StoreEntry(ICDClientInfo & aICDClientInfo) = 0;
+
+    /**
+     * Delete ICD Client persistent information associated with the specified scoped node Id.
+     * when ICD device is unpaired/removed, the corresponding entry in ICD storage is removed.
+     * @param aPeerNodeId scoped node with peer node id and fabric index
+     */
+    virtual CHIP_ERROR DeleteEntry(const ScopedNodeId & aPeerNodeId) = 0;
+
+    /**
+     * Remove all ICDClient persistent informations associated with the specified
+     * fabric index.  If no entries for the fabric index exist, this is a no-op
+     * and is considered successful.
+     * When the whole fabric is removed, all entries from persistent storage in current fabric index is removed.
+     *
+     * @param[in] fabricIndex the index of the fabric for which to remove ICDClient persistent information
+     */
+    virtual CHIP_ERROR DeleteAllEntries(FabricIndex aFabricIndex) = 0;
+
+    /**
+     * Validate received ICD Check-in message payload, return true when the key decription is successful
+     * Consumer has to provide keys and see whether it's the right key, further check its received
+     * counter value is valid, then return the correponding ICDClientInfo
```

Comment:
"then return the corresponding ICDClientInfo": need to express that ICDClientInfo is only updated on match.

Reply:
updated, thanks